### PR TITLE
feat: add Preview.js

### PR DESCRIPTION
**Pull request by fwouts**

### CODE REVIEW

**Comment by dominikg** [comment]
Hi, thank you for starting this.

The ci for this PR in ecosystem-ci only validates the integrity of ecosystem-ci itself, it doesn't test newly added tests.
please run `pnpm test previewjs` in this branch and check for errors.

Doing it locally i've seen 2 things that need improvement.
1) for some reason, turbo isn't found, i suggest you add the commands as scripts to your root package.json so it gets picked up correctly by pnpm run.
2) the packages in the `frameworks` subdirectory depend on other vite plugins. This is a very tricky situation, as plugins with different peer dependencies on vite could cause incompatibilities.

I currently have ho idea how this can be resolved unless the tests you want to add here don't use the frameworks section. For sveltekit which depends on vite-plugin-svelte, we actually run a build of vite-plugin-svelte first, but given that you have n frameworks, that sounds like a lot of work and easy breaks.

3) vite main branch is already on vite-4 alpha which has different types, if you see type issues, the suggested solution is to add a `vite-4` branch to your repo and update your test definition to dynamically select the branch in your repo that supports the same vite major

see https://github.com/vitejs/vite-ecosystem-ci/blob/main/tests/sveltekit.ts  for examples how to select branches dynamically and build other plugins to use in the tests.

Please ping us on vite discord #ecosystem-ci  in case you have questions.

**Comment by fwouts** [comment]
Hey @dominikg, thanks for the detailed feedback!

I realise now this was running locally for me because of my custom shell aliases, I'll add the root scripts as suggested. It also means this won't break if I ever migrate away from turborepo!

Thanks for the pointers for plugins. I think we could only run the tests for one specific framework (e.g. React) which would make it much more achievable to build the corresponding plugin as part of CI. It will also speed up CI.

I'll try and get this updated within a day or two 👍

**Comment by fwouts** [comment]
OK, it's now all green when running locally and I believe the issue with "turbo" not found should be resolved :) Let me know if there's anything else you'd like me to check!

Screen Shot 2022-11-28 at 10 45 48 am

File: tests/previewjs.ts
Comment (dominikg) [inline tests/previewjs.ts]: beforeTest can be used for additional setup, using it to run tests twice is very unusual. could it be removed or simplified?

Comment (fwouts) [reply tests/previewjs.ts]: This was a typo! I'm not sure why you're not seeing the latest version, maybe a GitHub bug?

Comment (patak-dev) [reply tests/previewjs.ts]: It looks fine for me:
```
		build: 'vite-ecosystem-ci:build',
		beforeTest: 'pnpm vite-ecosystem-ci:before-test',
		test: ['vite-ecosystem-ci:test']
```

Comment (dominikg) [reply tests/previewjs.ts]: not sure what happened, maybe a glitch in gh mobile ui. Command looks fine now, do you have to prefix it with pnpm though? build, beforeTest and test use 3 different styles for commands (script name, shell command, array with script name). If all of them are single scripts in package.json, would it hurt to use simple script name for all?

**Comment by dominikg** [comment]
still seeing an error locally:

```shell
@previewjs/plugin-react:e2e-test:   1) ../../../testing/dist/index.cjs:203:12 › smoke tests › nextjs-11 ==============================
@previewjs/plugin-react:e2e-test: 
@previewjs/plugin-react:e2e-test:     Test timeout of 120000ms exceeded.
@previewjs/plugin-react:e2e-test: 
@previewjs/plugin-react:e2e-test:     frame.waitForSelector: Target closed
@previewjs/plugin-react:e2e-test:     =========================== logs ===========================
@previewjs/plugin-react:e2e-test:     waiting for locator('#ready') to be visible
@previewjs/plugin-react:e2e-test:     ============================================================
@previewjs/plugin-react:e2e-test: 
@previewjs/plugin-react:e2e-test:        at ../../../chromeless/dist/index.cjs:107
```
(same for next-12 and next-13)

I also noticed turbo cache hits like `@previewjs/plugin-react:build: cache hit, replaying output 403eafd5e04e3191`
Is there remote caching in play or do you build it multiple times in this? Turbo caching is very finicky and due to the way vite-ecosystem-ci uses overrides and runs on a schedule it's extra work at best and replaying wrong results at worst, leading to very hard to track down errors.

Gotta discuss this with @patak-dev , but i think vite-ecosystem-ci should just set `TURBO_FORCE=true` to avoid it caching anything in all tests. (i wonder if we can also make it stop creating caches, just a wasted effort here imho)

**Comment by patak-dev** [comment]
Disabling caching sounds like a good idea. Also, if there are too many tests, @fwouts you may want to create a `test-vite` script that only runs a subset so we speed up Ecosystem CI. For several projects, we are not running their entire CI, only the tests that are directly helping make sure that your project and Vite work correctly.

File: tests/previewjs.ts
Comment (fwouts) [reply tests/previewjs.ts]: @dominikg you're right, for some reason I was under the impression that the `beforeTest` option had to be a shell command, but that's not the case. Fixed!

**Comment by fwouts** [comment]
Thank you both for reviewing!

Apologies for the failing smoke tests for nextjs, they're caused by a commit in the Preview.js repo I made today, after updating this PR. I guess that's the tricky thing with code that relies on a main branch that's consistently moving ahead :) I'll add `vite-ecosystem-ci:test` to Preview.js's own CI to make sure it doesn't happen in the future.

There isn't any remote caching happening, so a new run in CI would always be fresh unless we explicitly made the effort to cache `.turbo` directories (which we probably shouldn't). The reason why you're seeing a cache hit is that `vite-ecosystem-ci:test` runs `turbo test && turbo e2e-test`, which each implicitly run `turbo build` for the package and its dependencies, so there is overlap with the `vite-ecosystem-ci:build` script.

I'd prefer not to disable caching as I think it would then re-run the build multiple times and slow down the CI run. Alternatively I could change `vite-ecosystem-ci:test` to explicitly run the tests for each package but that seems a little convoluted, as there are quite a few of them (and that may change over time).

I'll also see how I can reduce the number of test, especially by only running tests for a single version of React (currently it's testing React 16, 17 an 18).

Quick todo list for myself:

- [ ] Fix Next.js smoke tests with Vite 4
- [ ] Add `vite-ecosystem-ci` workflow in Preview.js repo to make sure it never breaks
- [ ] Reduce number of tests to speed up CI (e.g. restrict to React 18 + smoke tests)

Cheers!

**Comment by dominikg** [comment]
fyi, turbo caching is already disabled on main of vite-ecosystem-ci, merge it in here and see what it does. 
ecosystem-ci is primarily useful for running tests that cover prevew.js use of vite apis. i don't imagine that changing between versions, so restricting the tests to a 1x1 matrix, maybe even just the smoketests would be great.

**Comment by dominikg** [comment]
also note that the vue and react plugins are going to move out of the main vite repo at some point, which makes my remarks about building them first relevant here too, so maybe try to stick to vanilla vite tests without additional frameworks first and then we can see how useful it would be to add the framework tests too.

File: tests/previewjs.ts
Comment (dominikg) [reply tests/previewjs.ts]: thanks

**Comment by fwouts** [comment]
> maybe even just the smoketests would be great.

The smoke tests would unfortunately not be sufficient to give me confidence, for example Preview.js relies heavily on [HMR events](https://vitejs.dev/guide/api-hmr.html#hot-on-event-cb) as well as [virtual modules](https://vitejs.dev/guide/api-plugin.html#virtual-modules-convention) which are some of the things I expect could break in subtle ways in a Vite update.

> also note that the vue and react plugins are going to move out of the main vite repo at some point, which makes my remarks about building them first relevant here too, so maybe try to stick to vanilla vite tests without additional frameworks first and then we can see how useful it would be to add the framework tests too.

That is good to know! Preview.js currently doesn't have a vanilla version so that's going to be difficult, but I'll see if I can come up with a new implementation for it. That will take a few weeks at least though. Happy to keep this PR on hold until then if you're not comfortable with the React dependency?

**Comment by dominikg** [comment]
hmr events and virtual modules both don't need react to be testable, could be done in a `npm create vite@latest --template vanilla` project. But i don't expect you to write new tests. If these are covered by your react tests currently then it is fine to run them, we just need to ensure that the react plugin is built to match the local vite we're using during the vite-ecosystem-ci run. 

This could mean doing a wrapped build like in sveltekit. It's possible this affects other cis too in the future, eg vite-plugin-ssr. @brillout, does your ci use the local vue or react plugins here?

**Comment by dominikg** [comment]
> > does your ci use the local vue or react plugins here
> 
> VPS's CI uses both `@vitejs/plugin-react` and `@vitejs/plugin-vue`. Not locally linked, but fetched from npm.

ecosystem-ci auto overrides these https://github.com/vitejs/vite-ecosystem-ci/blob/efa12aeef0f766949c6f0a7da8a6eda0e330dfb8/utils.ts#L218

**Comment by dominikg** [comment]
good news, plugin-react has recently moved into it's own repository and vite-ecosystem-ci needed a way to build it for other projects, so now you can just flag it and it's going to be built as a local override automatically.

https://github.com/vitejs/vite-ecosystem-ci/blob/main/tests/vite-plugin-ssr.ts#L9

**Comment by dominikg** [comment]
added that override and merged latest changes from main

**Comment by dominikg** [comment]
@fwouts can you check that `pnpm test previewjs` now runs your suite as expected and produces the correct result?

locally i saw the process hang after a while and it used 8 workers for react e2e tests, which can be a bit much for github ci. Does it use 50% of available threads?

**Comment by fwouts** [comment]
@dominikg I believe Playwright (which is used by Preview.js e2e tests) automatically uses as many workers as there are cores by default, but I've restricted it to 2 now in [the`vite-ecosystem-ci:test` script](https://github.com/fwouts/previewjs/blob/df5f62b9999e63ce486ac4bab79c66c255997230/package.json#L24).

I also ended up removing the build step and unit tests since disabling turbo caching resulted in builds being run multiple times (three times to be exact).

I ran the suite just now and unfortunately there's now one failing smoke test which looks due to a new breaking change in Vite 4. I guess I should get that fixed before we can merge this PR? I might not be able to get to that for a week or two unfortunately, hope you're OK with this PR being open for a little while longer.

**Comment by fwouts** [comment]
I should add that I had to update this line: https://github.com/vitejs/vite-ecosystem-ci/blob/5a64a2fa45bf759eceeee7a3d8bc612763950c55/utils.ts#L348

with:
```js
		fs.existsSync(v) ? `file:./${path.relative(dir, v)}` : v
```

I wonder if that's a Preview.js-specific issue?

**Comment by dominikg** [comment]
uhm what value of v was that needed for?  It should always use absolute paths for overriding so that they work correctly on different nesting levels inside a monorepo

**Comment by dominikg** [comment]
the error you observed with the absolute path override was a bug in pnpm 7.18.0, going to be fixed in 7.18.1

**Comment by fwouts** [comment]
> the error you observed with the absolute path override was a bug in pnpm 7.18.0, going to be fixed in 7.18.1

Great to know! I did find that a little odd, glad I didn't add it to this PR.

Separately, I'm going to see if I can upgrade Preview.js to Vite 4 beta in the next few days, I assume that will help with getting all tests to pass.

**Comment by fwouts** [comment]
Good news, I've upgraded Preview.js dependencies (including Vite 4!) and all tests now pass :)

Screen Shot 2022-12-10 at 12 27 50 am

This might be good to merge?

**Review by dominikg — APPROVED** [review]

**Pull request merged**
